File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/steps/TextWatcherEventInsertText.java
package org.wordpress.aztec.watchers.event.sequence.known.space.steps;

import org.wordpress.aztec.watchers.event.text.AfterTextChangedEventData;
import org.wordpress.aztec.watchers.event.text.BeforeTextChangedEventData;
import org.wordpress.aztec.watchers.event.text.OnTextChangedEventData;
import org.wordpress.aztec.watchers.event.text.TextWatcherEvent;


public class TextWatcherEventInsertText extends TextWatcherEvent {

    private CharSequence beforeText;
    private int insertionStart;
    private int insertionLength;

    public TextWatcherEventInsertText() {

    }

    private TextWatcherEventInsertText(BeforeTextChangedEventData beforeEventData, OnTextChangedEventData onEventData,
                                      AfterTextChangedEventData afterEventData) {
        this.beforeEventData = beforeEventData;
        this.onEventData = onEventData;
        this.afterEventData = afterEventData;
    }

    public void setInsertionStart(int insertionStart) {
        this.insertionStart = insertionStart;
    }

    public int getInsertionStart() {
        return insertionStart;
    }

    public int getInsertionLength() {
        return insertionLength;
    }

    public void setInsertionLength(int insertionLength) {
        this.insertionLength = insertionLength;
    }

    private boolean testBeforeTextChangedEventData(BeforeTextChangedEventData data) {
        beforeText = data.getTextBefore();
        return (data.getCount() == 0 &&  data.getAfter() > 0);
    }

    private boolean testOnTextChangedEventData(OnTextChangedEventData data) {
        return (data.getStart() >= 0 && data.getCount() > 0 && data.getTextOn().length() > 0);
    }

    private boolean testAfterTextChangedEventData(AfterTextChangedEventData data) {
        return beforeText.length() < data.getTextAfter().length();
    }

    @Override
    public boolean testFitsBeforeOnAndAfter() {
        return testBeforeTextChangedEventData(beforeEventData)
                && testOnTextChangedEventData(onEventData)
                && testAfterTextChangedEventData(afterEventData);
    }

    public static class Builder extends TextWatcherEvent.Builder {
        public TextWatcherEventInsertText build(){
            return new TextWatcherEventInsertText(beforeEventData, onEventData, afterEventData);
        }
    }
}

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/steps/TextWatcherEventInsertText.kt
package org.wordpress.aztec.watchers.event.sequence.known.space.steps

import org.wordpress.aztec.watchers.EndOfBufferMarkerAdder
import org.wordpress.aztec.watchers.event.text.AfterTextChangedEventData
import org.wordpress.aztec.watchers.event.text.BeforeTextChangedEventData
import org.wordpress.aztec.watchers.event.text.OnTextChangedEventData
import org.wordpress.aztec.watchers.event.text.TextWatcherEvent

class TextWatcherEventInsertText(beforeEventData: BeforeTextChangedEventData, onEventData: OnTextChangedEventData, afterEventData: AfterTextChangedEventData) : TextWatcherEvent(beforeEventData, onEventData, afterEventData) {

    private var beforeText: CharSequence? = null
    var insertionStart: Int = 0
    var insertionLength: Int = 0

    private fun testBeforeTextChangedEventData(data: BeforeTextChangedEventData): Boolean {
        beforeText = data.textBefore
        return data.count == 0 && data.after > 0
    }

    private fun testOnTextChangedEventData(data: OnTextChangedEventData): Boolean {
        return data.start >= 0 && data.count > 0 && data.textOn!!.length > 0
    }

    private fun testAfterTextChangedEventData(data: AfterTextChangedEventData): Boolean {
        return EndOfBufferMarkerAdder.safeLength(beforeText!!) < EndOfBufferMarkerAdder.safeLength(data.textAfter!!)
    }

    override fun testFitsBeforeOnAndAfter(): Boolean {
        return (testBeforeTextChangedEventData(beforeEventData)
                && testOnTextChangedEventData(onEventData)
                && testAfterTextChangedEventData(afterEventData))
    }

    class Builder : TextWatcherEvent.Builder() {
        override fun build(): TextWatcherEventInsertText {
            super.setGenericEventDataIfNotInit()
            return TextWatcherEventInsertText(beforeEventData, onEventData, afterEventData)
        }
    }
}
